Modules/Core/Sources/Core/JSON/JSONMapParser.swift

```diff
@@ -36,167 +36,161 @@ public enum JSONMapParseError : Error, CustomStringConvertible {
+        buffer.append(chunk)
     }
 
-    private var nextChar: Char {
-        return source[source.index(after: cur)]
+    private var bytesInBuffer: Int {
+        return buffer.count
     }
 
-    private var distanceToEnd: ByteSequence.IndexDistance {
-        return source.distance(from: cur, to: end)
+    private func getCurrentByte() throws -> Byte {
+        if bytesInBuffer == 0 {
+            try readChunk()
+        }
+        guard bytesInBuffer >= 1 else {
+            throw insufficientTokenError(reason: "unexpected end of tokens")
+        }
+        return buffer[0]
     }
 
-    private var currentSymbol: Character {
-        return Character(UnicodeScalar(currentChar))
+    private func getCurrentSymbol() throws -> Character {
+        return try Character(UnicodeScalar(getCurrentByte()))
     }
 
     private func parseSymbol(_ target: StaticString, _ iftrue: @autoclosure (Void) -> Map) throws -> Map {
-        if expect(target) {
+        if try expect(target) {
             return iftrue()
         }
-        throw unexpectedTokenError(reason: "expected \"\(target)\" but \(currentSymbol)")
+        let symbol = try getCurrentSymbol()
+        throw unexpectedTokenError(reason: "expected \"\(target)\" but got \(symbol)")
     }
 
     private func parseString() throws -> Map {
-        advance()
+        try advance()
 
         var buffer = [CChar]()
 
-        while cur != end && currentChar != Char(ascii: "\"") {
-            switch currentChar {
-            case Char(ascii: "\\"):
-                advance()
-
-                guard cur != end else {
-                    throw invalidStringError(reason: "missing double quote")
-                }
-
-                guard let escapedChar = parseEscapedChar() else {
-                    throw invalidStringError(reason: "missing double quote")
+        while try getCurrentByte() != Byte(ascii: "\"") {
+            switch try getCurrentByte() {
+            case Byte(ascii: "\\"):
+                guard let escapedChar = try parseEscapedCharacter() else {
+                    throw invalidStringError(reason: "invalid escaped character")
                 }
 
                 String(escapedChar).utf8.forEach {
                     buffer.append(CChar(bitPattern: $0))
                 }
             default:
-                buffer.append(CChar(bitPattern: currentChar))
+                buffer.append(CChar(bitPattern: try getCurrentByte()))
             }
-
-            advance()
+            try advance()
         }
 
-        guard expect("\"") else {
+        guard try expect("\"") else {
             throw invalidStringError(reason: "missing double quote")
         }
-
+        
         buffer.append(0) // trailing nul
         return .string(String(cString: buffer))
     }
 
-    private func parseEscapedChar() -> UnicodeScalar? {
-        let character = UnicodeScalar(currentChar)
+    private func parseEscapedCharacter() throws -> UnicodeScalar? {
+        try advance()
```

Comment:
is this here on purpose? (if so, could you add a comment explaining what its for)